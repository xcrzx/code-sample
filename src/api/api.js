import axios from 'axios'

const api = axios.create({
  baseURL: 'http://localhost:9001/api/',
})

const getTodos = () => api.get('/todos/')

export default {
  getTodos,
}
