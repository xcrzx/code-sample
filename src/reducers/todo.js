import { handleActions } from 'redux-actions'
import R from 'ramda'

import ActionTypes from 'constants/ActionTypes'
import { setIsFetching, unsetIsFetching, mergePayload } from './common'

export const defaultState = {
  isFetching: false,
  entities: {},
}

const todoReducer = handleActions({
  [ActionTypes.todo.fetch.request]: setIsFetching,
  [ActionTypes.todo.fetch.response]: {
    next: R.pipe(mergePayload, unsetIsFetching),
    throw: unsetIsFetching,
  },
}, defaultState)

export default todoReducer
