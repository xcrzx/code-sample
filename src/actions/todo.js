import { createAction } from 'redux-actions'
import R from 'ramda'

import ActionTypes from 'constants/ActionTypes'
import api from 'api/api'

export const toggleTodo = createAction(ActionTypes.todo.toggle)

export const fetchTodosRequest = createAction(ActionTypes.todo.fetch.request)

export const fetchTodosResponse = createAction(ActionTypes.todo.fetch.response)

export const fetchTodos = () => (dispatch) => {
  dispatch(fetchTodosRequest())

  return api.getTodos()
    .then(R.pipe(R.prop('data'), fetchTodosResponse, dispatch))
    .catch(R.pipe(R.path(['error', 'message']), R.construct(Error), fetchTodosResponse, dispatch))
}
