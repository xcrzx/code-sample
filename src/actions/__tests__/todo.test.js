import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'

import api from 'api/api'
import { fetchTodos, toggleTodo } from 'actions/todo'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe('Todo actions', () => {
  test('fetchTodos success', () => {
    api.getTodos = jest.fn(() => Promise.resolve({ data: { entities: {} } }))

    const store = mockStore({ todo: {} })

    return store.dispatch(fetchTodos())
      .then(() => expect(store.getActions()).toMatchSnapshot())
  })

  test('fetchTodos error', () => {
    api.getTodos = jest.fn(() => Promise.reject('Something went wrong'))

    const store = mockStore({ todo: {} })

    return store.dispatch(fetchTodos())
      .then(() => expect(store.getActions()).toMatchSnapshot())
  })

  test('toggleTodo', () => {
    const store = mockStore({ todo: {} })

    store.dispatch(toggleTodo(0))
    expect(store.getActions()).toMatchSnapshot()
  })
})
