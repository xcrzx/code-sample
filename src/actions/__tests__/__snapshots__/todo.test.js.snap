// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo actions fetchTodos error 1`] = `
Array [
  Object {
    "type": "TODO_FETCH_REQUEST",
  },
  Object {
    "error": true,
    "payload": [Error],
    "type": "TODO_FETCH_RESPONSE",
  },
]
`;

exports[`Todo actions fetchTodos success 1`] = `
Array [
  Object {
    "type": "TODO_FETCH_REQUEST",
  },
  Object {
    "payload": Object {
      "entities": Object {},
    },
    "type": "TODO_FETCH_RESPONSE",
  },
]
`;

exports[`Todo actions toggleTodo 1`] = `
Array [
  Object {
    "payload": 0,
    "type": "TODO_TOGGLE",
  },
]
`;
